### PR TITLE
Delete expired payments

### PR DESCRIPTION
**Pull request by remyers**
This is the first step to bring automatic purging of expired incoming payments to projects built on lightning-kmp. This ability was requested in issue [#2078](https://github.com/ACINQ/eclair/issues/2078)  in Eclair.

This PR adds functions to find and remove expired incoming payments similar to Eclair PR [#2174](https://github.com/ACINQ/eclair/pull/2174). This PR does not launch an automatic thread for purging expired invoices like in Eclair because mobile implementations may not always be running and may want to purge invoices using a platform specific trigger.

Additional tasks are needed to bring this to Phoenix, including:
[  ] add  platform specific database support for new functions added to `InMemoryPaymentsDb` eg. [SqlitePaymentsDb.kt](https://github.com/ACINQ/phoenix/blob/master/phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/db/SqlitePaymentsDb.kt)
[  ] add platform specific tests
[  ] add platform specific triggers to call PurgeExpiredPayments

### CODE REVIEW

**Comment by remyers** [comment]
@t-bast - a few questions:
* is it the correct approach to defer automatic purging to platform specific app implementations instead of in the lightning-kmp library?
* OK to continue with next step as PRs to the Phoenix repo before this is merged?

**Comment by t-bast** [comment]
> is it the correct approach to defer automatic purging to platform specific app implementations instead of in the lightning-kmp library?

Yes, I think it makes sense to let the application call us to purge expired invoices. Let's first add that capability and then we'll see with @dpad85 how Phoenix will use it.

> OK to continue with next steps as PRs to the Phoenix repo before this is merged?

I'm not sure I fully understand what you ask, but I think you should simply add the ability to purge expired invoices on-demand to lightning-kmp, and we'll let @dpad85 and @robbiehanson handle the Phoenix side and call it.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:51]: I'm not sure we should add this function, I don't see it being used by wallets.
Users are only interested in listing the payments that they received.
@dpad85 let us know what you think.

File: src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Applications building on top of `lightning-kmp` won't be able to call this function right now, as the payment handler is a private field in a `Peer` instance.

Can you create a new [payment event](https://github.com/ACINQ/lightning-kmp/blob/234e049f7ee93296cfa03caa0e9de7b1f2fe33ae/src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt#L41), for example `data class PurgeExpiredPayments(val fromCreatedAt: Long, val toCreatedAt: Long) : PaymentEvent()` and add a handler for it in the `processEvent` function?
That will let Phoenix easily interact with this feature.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:57]: What about returning a boolean here to indicate whether the payment was removed or not?
This would let us return `false` for paid (or partially paid) payments.

File: src/commonTest/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandlerTestsCommon.kt
Comment (t-bast) [inline src/commonTest/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandlerTestsCommon.kt:1170]: I'd like to avoid introducing delays in unit tests, otherwise that tends to creep in many places and we end up with a very slow test runs. We really want our test suite to run as fast as it can.

We can achieve by creating the received payment directly in the DB instead of going through `paymentHandler.process`:

```kotlin
val db = InMemoryPaymentsDb()
val paymentHandler = IncomingPaymentHandler(TestConstants.Bob.nodeParams, TestConstants.Bob.walletParams, db)

// create incoming payment that has expired and not been paid
val expiredInvoice = paymentHandler.createInvoice(randomBytes32(), defaultAmount, "expired", listOf(), expirySeconds = 3600,
    timestampSeconds = currentTimestampSeconds() - 3600 - 60)

// create incoming payment that has expired, but has been paid before its expiry
val paidInvoice = paymentHandler.createInvoice(defaultPreimage, defaultAmount, "paid", listOf(), expirySeconds = 3600,
    timestampSeconds = currentTimestampSeconds() - 3600 - 60)
db.receivePayment(defaultPaymentHash, setOf(IncomingPayment.ReceivedWith.LightningPayment(defaultAmount, randomBytes32(), 3)), receivedAt = currentTimestampSeconds() - 1800)
```

**Review by t-bast — COMMENTED** [review]
Looks mostly good, just a few improvements to consider.

File: src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Ah ha, I didn't think of that. Should PurgeExpiredPayments notify the calling application by sending a PeerListenerEvent like this then?
`object PurgeExpiredPaymentsCompleted : PeerListenerEvent()` 

(assuming I change to return a boolean and not number of purged invoices)

Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Good question, I don't know if the application is really interested about knowing that it has completed.
I think they only want to trigger this as a fire-and-forget task, but @dpad85 or @robbiehanson will know better.
I'd say that for now we don't emit an event when it completes, and we can add it if they think it's useful?

Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Actually shouldn't be Boolean message, but can return the number of successfully purged expired invoices :
data class PurgeExpiredPaymentsCompleted(val count: Int) : PeerListenerEvent()

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt]: fixed in [be38964](https://github.com/ACINQ/lightning-kmp/pull/322/commits/be38964082f80475adeb7ff0f2bd9455988f5afb); I also only count successfully removed payments in the `IncomingPaymentsHandler`.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:51]: I added it to be extra careful in the test and check the exact state of the DB before and, more importantly, after purging invoices.

If it's only useful for testing so perhaps I should take it out. I can instead use the fact that only one incoming payment is purged, but not two, to ensure that partially paid payments aren't purged.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt]: On second thought, `listExpiredPayments` will not return paid (or partially paid) incoming payments so it should be a rare race condition to have on that list expired payments that were [partially] paid. It is still worth keeping this change though for the case where `removeIncomingPayment` is used for something where an absent payment is more likely (or important).

File: src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt
Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Added `PurgeExpiredPayments` in [53a9360](https://github.com/ACINQ/lightning-kmp/pull/322/commits/53a9360cad36172ef6123ba8b592e62500a18da4) but I will hold off on a return message until hearing what @dpad85 and @robbiehanson think.

Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Btw, what is the policy towards removing warnings in a file I changed that weren't caused by my changes? They could just be caused by differences in linter settings, but there were a few in Peer.kt .

File: src/commonTest/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandlerTestsCommon.kt
Comment (remyers) [reply src/commonTest/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandlerTestsCommon.kt]: Nice! It's easier to read this way too. Fixed in [abd5760](https://github.com/ACINQ/lightning-kmp/pull/322/commits/abd576080cd4b240c509346e83efbb90ac2b6a71).

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:51]: Let's wait for dpad85 to decide!

File: src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: There are a few about fields that could be private, but I think those shouldn't be fixed because Phoenix actually uses those fields so it needs to have access to it (maybe there's a kotlin tag to tell the compiler that these fields really need to stay public?). Apart from those, I think it's always good to fix a few warnings when you see them, even if they're unrelated to your change ;)

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt]: Agreed, we will probably never see it return `false` because we're using it correctly, but it makes sense to keep that boolean for people who may use this function differently in the future.

File: src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt
Comment (t-bast) [inline src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:394]: I think you want to use `filter` instead of `map` to count only the ones that actually get removed?

Comment (remyers) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt]: Doh! yes, actually it seems `count` applied to `removeIncomingPayment` as a predicate is what I really wanted here. Fixed in [290d8df](https://github.com/ACINQ/lightning-kmp/pull/322/commits/290d8dfcd3afb500aee2bbea91e06a14a816ca91).

**Review by t-bast — APPROVED** [review]
LGTM, well done! Let's wait for @dpad85 or @robbiehanson to chime in on some of the comments, feel free to merge when they're satisfied.

File: src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt
Comment (dpad85) [reply src/commonMain/kotlin/fr/acinq/lightning/payment/IncomingPaymentHandler.kt:393]: Regarding the return type of `purgeExpiredPayments`, we don't need that information on Phoenix's side. We do not even need to know it succeeded.

Also, since in Phoenix the expiration information we need is [actually stored as JSON in the `incoming_payments` database](https://github.com/ACINQ/phoenix/blob/19c5f252ef2be7e107f989d1ab034b3208a4ef79/phoenix-shared/src/commonMain/paymentsdb/fr.acinq.phoenix.db/IncomingPayments.sq#L11), we won't be able to make a single SQL query that would delete things in one go. Instead, we'll have to make one SQL query for each payments in the given range, go back to kotlin, deserizalize the json data, check the expiration status and then actually make a new SQL query to delete that payment. So we can expect this operation to be a quite slow! 

Btw, I note that in our implementation of the query for expired payments, we should first check that the `received` rows are null for better perfs.

File: src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt
Comment (dpad85) [reply src/commonMain/kotlin/fr/acinq/lightning/db/PaymentsDb.kt:51]: Indeed we don't need it in the app, but it's nice to have that check in the tests. So I think we should keep it, especially if `purgeExpiredPayments` does not return the affected rows count anymore (which affects the test).

**Comment by robbiehanson** [comment]
I think this makes some sense from an interface perspective. That is, when building atop of lightning-kmp, the developer will be asked to implement functions to search for expired payments. Which is a good reminder that this task should be performed.





The actual implementation of this task will be highly dependent upon how they structure the database layer. In the case of Phoenix, we need to:

- check that `received_at` is null
- deserialize the `origin_blob` to see if the payment is expired



So what we will *probably* do is something like:

- add an index on `receieved_at`
- fetch a batch (10?) of rows where `received_at` is null, sorted by `created_at`
- delete the expired items
- stop if the batch was empty, or any were not expired
- repeat as needed, each iteration using a different database transaction



This means we might not implement `listExpiredPayments` in Phoenix, instead opting for an optimized garbageCollection function.

**Pull request merged**
